library/asn1parse.c

```diff
@@ -150,12 +150,28 @@ int mbedtls_asn1_get_int( unsigned char **p,
     int ret;
     size_t len;
 
+#if defined(MBEDTLS_ASN1_PARSE_NEGINT)
+    unsigned char *neg = 0;
+#endif /* MBEDTLS_ASN1_PARSE_NEGINT */
+
     if( ( ret = mbedtls_asn1_get_tag( p, end, &len, MBEDTLS_ASN1_INTEGER ) ) != 0 )
         return( ret );
 
-    if( len == 0 || len > sizeof( int ) || ( **p & 0x80 ) != 0 )
+    if( len == 0 || len > sizeof( int ) )
         return( MBEDTLS_ERR_ASN1_INVALID_LENGTH );
 
+    if ((**p & 0x80) != 0)
+#if defined(MBEDTLS_ASN1_PARSE_NEGINT)
+    {
+        neg = *p;
+        **p &= ~(0x80);
+    }
+#else
+    {
+        return( MBEDTLS_ERR_ASN1_INVALID_DATA );
```

Comment:
@gilles-peskine-arm @mpg What's your opinion on this? Should we return `INVALID_DATA`, assuming that whenever we call the parsing function in a compilation where `ASN1_PARSE_NEGINT` is unset, it is implicit that negative numbers are not acceptable, or should we rather return something like `UNSUPPORTED_FEATURE`? I'm inclined to the latter, because it doesn't mix the ASN.1 parsing with the context-dependent judgement of whether negative numbers are allowed or not.